### DISCUSSION TITLE
Make Ulauncher available through Flatpak and Flathub

### DISCUSSION THREAD

**Discussion by rejurime** [Ulauncher/Ulauncher#883] [Packaging]
It would be great to have Ulauncher available through Flatpak. This would make it easier to distribute Ulauncher.

---

If someone could try doing this manually and then listed the steps here with any caveats found along the way, I (@gornostal) will automate the releases for Flatpak

**Comment by okseb** [comment]
I would be interested in seeing this as well.

**Comment by gornostal** [comment]
I'm not sure how I feel about this.
Yes, it would be good to have a flatpack package, as well as snap (#208), but I don't want to drop deb and rpm packages.

That means more testing, more maintenance work, and probably just a little bit more new users :thinking:

**Comment by rprihoda** [comment]
Well if you plan to go only deb and rpm packages then you'd need/should  setup repos for each distro, which is additional maintenance on it's own. If you go flatpak/snap you no longer have to worry about maintaining repos for each distro. 

Just my opinion.

**Comment by pierrehenrymuller** [comment]
Personally I prefer deb packages, in companies with secure workstations, it is easier to offer a deb to administrators who will deploy them on workstations than a flatpack or snap that are not accepted because it would allow users to install their own applications without being administrators.

If you are going in the direction of packaged applications, also keep the deb and rpm for those who have no choice. I chose Ulauncher also on this aspect.

**Comment by rakshazi** [comment]
Hello from 2020!

Flatpak is still a good thing, so add a flatpak distribution will be a huge plus

**Comment by owzim** [comment]
Anyone can create a flatpak version of ulauncher. It does not have to be the developers themselves. If you want flatpak, try researching how to build it and then contribute. @gornostal is right, this adds maintenance overhead, so any help is certainly appreciated. /cc @rakshazi @rowbawts @rejurime 

ps: I do not use ulauncher anymore, just a bystander :)

**Comment by ghost** [comment]
@gornostal Would you plese make this available in Flatpak? I am dying to run it on Centos 7. Please please please...

**Comment by gornostal** [comment]
@bwv656 I've marked this issue as contributor-friendly. 
Looking for people who could give clear steps that they've tried and that work on how create a Flatpak build for ulauncher. Then I'm going to add that to the release automation.

**Comment by iCSpotRun** [comment]
I understand Flatpak's to have privacy and security benefits over anything depending on PPA's. If I'm right, wouldn't this be motivation enough to support Flatpak's? If not, please kindly educate this Noob ;)

**Comment by gornostal** [comment]
>  ... over anything depending on PPA's

Some user reported (tagged @Ulauncher in twitter) that Ubuntu marks my PPA as not secure. So I would agree with you on that.
I may have not configured it correctly :thinking: 

Then to your point about Flatpak's security, I got curious about it and quickly stumbled upon this website: https://flatkill.org/
Haven't read it fully yet. But even if Flatpak's security is a myth, I see benefits in having a flatpak for Ulauncher.
It's just not something I would now focus myself, but I would support contributors.

**Comment by nandgator** [comment]
Hi, this [Reddit  post](https://www.reddit.com/r/linux/comments/9n50ba/lets_see_why_flatpak_and_sandboxing_are_awesome/) has a response to https://flatkill.org.

Flatpak does not have cli support (not till now) but snap does. So will packaging them in these two formats require much change in the source code?

By the way if anyone's interested, here's the official docs to build [flatpak](https://docs.flatpak.org/en/latest/) and [snap](https://snapcraft.io/build) packages.

**Comment by friday** [comment]
The package file would be hundreds of megabytes instead of 7mb btw. We have some big dependencies (GTK and WebkitGTK).

Would we need to add conditions to our code for this, to access the host file system (like scanning for desktop entries, using system icons, picking icons for shortcuts) and launch app with the host file system scope, access to system themes, and not be killed if ulauncher crashes or is terminated (#778)?

How about extensions that access the file system? For example my clipboard extension uses the system themes `edit-paste` icon, and it integrates with three different clipboard managers by running shell commands. It uses libnotify to show notifications on the host system and it even supports running a custom command to paste https://github.com/friday/ulauncher-clipboard#copy-hook. It's not exactly sandbox-friendly.

And would we be able to have a config/data directory that persists if people uninstall and reinstall?

**Comment by gornostal** [comment]
> It's not exactly sandbox-friendly

I'll second that!
I recently looked into Flatpak docs, the sandbox permissions have to be defined pre-build, which makes it difficult (even impossible) to allow users to change the sandboxing scope at runtime.

**Comment by tim77** [comment]
> The package file would be hundreds of megabytes instead of 7mb btw.

No, it will not.

> We have some big dependencies (GTK and WebkitGTK).

This is part of GNOME runtime https://docs.flatpak.org/en/latest/available-runtimes.html#gnome

**Comment by friday** [comment]
> This is part of GNOME runtime https://docs.flatpak.org/en/latest/available-runtimes.html#gnome

That's nice. It has a lot of the heavy libraries we use actually. WebKitGTK, PyGObject, Libnotify and GObject Introspection  :+1: 

Open Build Service added support for flatpak this year, so if we migrate to that (which we have several others reasons to want to do) supporting Flatpak might not require a lot of additional work for us. Or it might. We have a different focus for now before we can make time to look  at packaging.

**Comment by Arcitec** [comment]
I'd say that Flatpak has become the premier distribution method for Linux desktops now. GNOME and KDE agree. They're both integrating Flathub's app donation/payment features too. And Flatpaks are the main way to install applications on immutable distros.

I realize that nobody on the team uses any of this yet, but maybe someday it'll be worth it to create an official Flathub release. You'd definitely start earning donations from it. I'm one of the people who is just waiting for Flathub donations to become available so that I can start throwing money at all my favorite app developers.

**Reply by Arcitec** [reply]
As for solving the issue of how to run a background task permanently (the launcher/listener): That's a solved problem. The "Pika Backup" flatpak registers itself for Background Activity and registers a `~/.config/autostart/org.gnome.World.PikaBackup.desktop` which runs the following permanent background process:

```sh
[Desktop Entry]
Type=Application
Name=org.gnome.World.PikaBackup
Exec=flatpak run --command=pika-backup-monitor org.gnome.World.PikaBackup
X-Flatpak=org.gnome.World.PikaBackup
```

Any other needs Ulauncher might have are fully supported too, such as DBus services, DBus activation (instead of .desktop `Exec=` activation), "flatpak portals" for all things such as keybindings etc.

**Reply by friday** [reply]
Flatpak evangelists always say it is easy/possible, but I haven't seen any PRs or unofficial builds for any desktop application launcher or extensible plugin system capable of running arbitrary code on your system, and Ulauncher is both. See https://github.com/albertlauncher/albert/issues/407 and https://github.com/albertlauncher/albert/commit/d223659865caae971fb718f2b400aa4429235a53 for example.

**Reply by Arcitec** [reply]
@friday Specifically Albert or Ulauncher, you're right, those don't exist on Flathub. But others do, such as https://flathub.org/apps/hu.kramo.Cartridges

It can launch *any* application or command on the host. It achieves it with one small change to the executed command:

`flatpak-spawn --host <put command here>`

This command is available to Flatpaks that have the following in [their manifest](https://github.com/flathub/hu.kramo.Cartridges/blob/master/hu.kramo.Cartridges.json):

`"--talk-name=org.freedesktop.Flatpak",`

(Source: [flatpak-spawn](https://manpages.ubuntu.com/manpages/kinetic/man1/flatpak-spawn.1.html))

There are plenty of other examples, such as [Black Box](https://flathub.org/apps/com.raggesilver.BlackBox) which is a terminal GUI that uses [the same permission](https://github.com/flathub/com.raggesilver.BlackBox/blob/master/com.raggesilver.BlackBox.json) to launch a terminal that runs completely unsandboxed on the host. The sandboxed application is able to connect to regular stdin/stdout pipes of the `flatpak-spawn`'d process.

Likewise, persistence/background processes/"autostart apps" are doable with the same method as Pika Backup mentioned above.

Plugin frameworks would work as-is without any changes necessary, because they'd run inside/side by side with Ulauncher inside the sandbox. The only issue with Plugins would be the ones that try to access the host filesystem, but that's solvable with one neat little trick: `--filesystem=host`, which means the Flatpak application is able to read/write any file it wants to on the host.

So let's recap:

- Flatpak-Spawn Host: Runs any command outside the sandbox. So the only change needed to Ulauncher's "launch something" code is a "if inside flatpak, prepend `flatpak-spawn --host`".
- Host filesystem permissions: With this flag, it can access any files on the host.
- Background/autostart app: This is doable too, via the same method as Pika Backup above.

I don't think it's worth it for you to try to do this though. Maybe later, when Flathub has the 1-click donation system that's fully integrated into KDE Discover and GNOME Software (later this year/next year), then it might be worth it. :smile_cat:

**Reply by otsaloma** [reply]
> Host filesystem permissions: With this flag, it can access any files on the host.

The most common thing launchers do is index apps. I tried making a flatpak of [Catapult](https://github.com/otsaloma/catapult) long ago and had an "oh, of course" moment when I noticed it was just listing apps in the sandbox. Just checked again as I noticed that in the sandbox certain host directories, like `/usr` are mounted under `/var/run/host`. I tried setting `$XDG_DATA_DIRS`, but I couldn't get `Gio.AppInfo.get_all()` to list my host apps. Interestingly it did list a couple host apps I have installed under `/opt` with just desktop files under `/usr/share/applications`, but not anything installed under `/usr` – I assume there's some kind of a conflict there. So, yes, technically you can access the host filesystem, but it looks still far from being usable (if not wanting to write a bunch of custom flatpak-specific code).

https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access

**Reply by ManuelSchneid3r** [reply]
tl;dr For launchers flatpak is nothing less than an _absolute catastrophe_.

@otsaloma it is hefty to get launchers into flats. on matrix i got told flatpak is not for us.

dealbreakers:
* / (root) is mapped to /run/host, which has a tremendous effect on most plugins   ( /bin, /usr/share, …)
* PATH, ofc it is different if the entire filesystem hierarchy got remapped
* XDG_*_DIRS. Desktop entry spec host apps? It maay be possible to write a wrapper script that queryies the host env using flatpak-spawn and try to inject is somehow into the sandboxed app (not sure if overwriting the overwritten env is a good idea. i mean i think it is there for a reason)
* XDG_*_HOME cache, config, data, ipc sockets - everything. in. different. places.
* Running processes: Yeah one could try to remap any started commandline path. Fortunately at albert most calls go through albert:: runDetachedProcess(…), but is this sophisticated? Any plugin using POpen or QProcess will end up in the sandbox or have to take care of remapping paths on its own.
* …

In the end the transition would be far from minimal invasive. A thing I also absolutely dislike about wayland. Seriously, since about 10 years or so development on linux got a major pain. Let alone the distribution problematic. Finally Flatpak to the rescue I thought. Oh dear… 

I spend a lot of time on flatpak. and i think i should stop sooner than later  or at least give canonicals snapcraft a shot before. by the way OBS is not a perfect solution either (it could, but it is not atm).

**Reply by Arcitec** [reply]
@ManuelSchneid3r All of that is solvable by having Ulauncher launch a helper process that runs on the host.

Example:

- `flatpak-spawn --host python <helper script>`
- Connect stdin and stdout of that process to talk to it securely.
- Then Ulauncher can run its scanning, app launching code on the host, in the helper process. Which funnels information back into the sandboxed UI.

Lots of other apps use this kind of approach. For example, the Ptyxis terminal emulator spawns a permanent helper process on the host to watch running processes.

Even the non-sandboxed Ulauncher would benefit from this layout (a separate helper process) because it would make the app more stable.

But again, I don't think it's worth all the hassle to refactor the app, at least not until immutable distros like Silverblue are used by most people.

**Reply by ManuelSchneid3r** [reply]
Sure but I want to write an app. Not a driver.

**Reply by otsaloma** [reply]
> All of that is solvable by having Ulauncher launch a helper process that runs on the host.

And then users need to install dependencies of that script manually on the host?

In case of launchers that can be e.g. `python3-gi` and `gir1.2-glib-2.0` (names from debian).

**Reply by friday** [reply]
> Even the non-sandboxed Ulauncher would benefit from this layout (a separate helper process) because it would make the app more stable.

I don't disagree. I have [said the same thing some years ago](https://github.com/friday/Legacy-Ulauncher-Fork/discussions/22), before I was given the maintainer role, and before Pop-os launcher that went this route (although with a daemon).

This would require some work though, and it's not something I'm looking at now. Additionally if you have two separate dependencies with separate sub-dependencies, one which can be installed as a flatpak and the other can't be, then I would say that makes it harder to install Ulauncher, and this defeats the purpose of going flatpak doesn't it?

**Reply by friday** [reply]
> Sure but I want to write an app. Not a driver.

I don't know about that. All the issues and limitations are related to the app. I would be happier maintaining a launcher-service rather than a launcher app. But you your options isn't app or driver/service. It's app vs app + driver/service.

**Reply by ManuelSchneid3r** [reply]
The need for this workaround is conceptual nonsense imho. The thing I dislike about Flatpak is that it forces authors to write code required by the sandbox. If you use the sandbox or not, but especially if you don't use it. It is never going to be the future of software distribution, _because_ it forces you to write platform dependent adoption code (the driver). It's a pity. I really was excited about the idea of uniform software distribution.

**Reply by ManuelSchneid3r** [reply]
If anybody adopts anything then this should be flatpak allowing authors to simply opt out the sandbox and just use the LD_LIBRARY paths of the flatpak runtime.

**Comment by ddouglas87** [comment]
If Flatpak is not going to work, what about Snap?  In 2024, snaps are more wildly available than Flatpaks today.  Snap is supported on equal or more operating systems than Flatpak is, and snap has a much larger repository of applications for users.  That and Snap has sandboxing options keeping the user secure similar to Flatpak but without the overhead and unlike with Flatpak, snap's sandboxing options are optional.  This is perfect for Ulauncher as it probably can not run properly in a sandboxed environment.

I'm on OpenSUSE.  The only option I have right now for auto updated versions are relying on Dead_Moody's 3rd party OPI.  This is risky because this person can put a virus in one day and I'd be none-the-wiser.  Today all of the 3rd party software I install recommends Snap (except Ulauncher and Chrome).  Snap works and works well for me.  As an end user it works better than Flatpak (though I don't mind using Flatpak if that's what the developer wants).

I get logically why Flatpak is ideal, but it's such a pain in the ass that most developers default to supporting Snap instead.  Why not Ulauncher too?  Snap will provide support to I believe all major Linux operating systems.

**Reply by friday** [reply]
I'm not against it, but this won't be a priority for me personally. If someone else does the research and work and contribute a PR this is a lot more likely to happen.

**Comment by arigit** [comment]
+1 on this. 

Very long time uLauncher user, this is the last important desktop application that I have not yet been able to run on flatpak. I am aware in the beginning the transition to desktop containers was difficult for many apps but today we have Terminals, entire multimedia centers running on GBM/Wayland or X11, LibreOffice, browsers, development environments, all kinds of tools - you name it. 

All may other desktop day to day applications are in flatpaks as of now - minus ulauncher. As a consequence - upgrading my Desktop PC host OS after 3 years has been the most painless upgrade I've done in 25 years using desktop linux. None of my desktop apps even noticed it, amazing considering I switched from X11 to Wayland in the process. That alone tells a story of how the linux SW ecosystem matured and evolved.

Please prioritize flatpaks over snaps as the ubuntu-only approach to desktop containers will limit the effort to a much narrower user base

**Comment by kmf** [comment]
Everyday I'm amazed at the pure amount of packages available through Flathub.
It's the quickest way for people to try your application and most effective way of distribution a Desktop Application to Linux users.

Ignoring this, will be a missed opportunity.

**Comment by friday** [comment]
Any more comments about how we need/should be doing this will be marked as off topic. Just use the up vote / thumbs up reaction on the original post if you think this is important for you but don't have anything to add about the technical implementation / difficulties.

We need people to help. Opinions alone won't help.

**Comment by pierrehenrymuller** [comment]
For the information of you, companies that deploy Linux as a workstation block users from installing applications.
This is achieved by :
- not being an administrator (so as not to install with deb, rpm, etc.)
- using the noexec option on /home (to prevent appimages)
- uninstalling flatpack or snapd managers (to counter these packages)
- forced use of dedicated repo to ensure updates.

Also, when a user wishes to have an application, the only solution after their IT department has approved it is to use deb or rpm.

I've currently had to stop using Ulauncher for this reason alone, and have switched to Albert, which provides deb and rpm.

There are several ways of installing software under Linux, and you have to think about all of them in the same way as under Windows or MacOS:
- the user has his own computer and is an admin: he can use any installation method he likes
- the user is not an admin but has flatpack / snapd available: he'll take this option
- if the user can't install anything, he'll have to ask his IT department: the only solution is to use packages from Linux distros with a repo associated with the project
